src/main/java/org/opensearch/knn/plugin/transport/UpdateModelGraveyardTransportAction.java

```diff
@@ -156,4 +168,72 @@ public ClusterTasksResult<UpdateModelGraveyardTask> execute(ClusterState cluster
             return new ClusterTasksResult.Builder<UpdateModelGraveyardTask>().successes(taskList).build(updatedClusterState);
         }
     }
+
+    private static void checkIfIndicesAreUsingModel(ClusterState clusterState, UpdateModelGraveyardTask task, String[] indicesArray)
+        throws IOException {
+        Map<String, MappingMetadata> mappings = clusterState.metadata()
+            .findMappings(indicesArray, task.getIndicesService().getFieldFilter());
+        Map<String, IndexMetadata> indices = clusterState.metadata().indices();
+        List<String> indicesUsingModel = new ArrayList<>();
+        // Parse indices and add to list if using the model
+        for (Map.Entry<String, MappingMetadata> entry : mappings.entrySet()) {
+            MappingMetadata mappingMetadata = entry.getValue();
+            Settings indexSettings = indices.get(entry.getKey()).getSettings();
+            if (mappingMetadata != null && Boolean.parseBoolean(indexSettings.get("index.knn", "false"))) {
+                indicesUsingModel = parseMappingMetadata(mappingMetadata, task, indicesUsingModel, entry.getKey());
+            }
+        }
+        // Throw exception if any indices are using the model
+        if (!indicesUsingModel.isEmpty()) {
+            throw new DeleteModelException(
+                String.format(
+                    "Cannot delete model [%s].  Model is in use by the following indices %s, which must be deleted first.",
+                    task.getModelId(),
+                    indicesUsingModel
+                )
+            );
+        }
+    }
+
+    private static List<String> parseMappingMetadata(
+        MappingMetadata mappingMetadata,
+        UpdateModelGraveyardTask task,
+        List<String> indicesUsingModel,
+        String index
+    ) {
+        Map<String, Object> mappingMetadataSourceMap = mappingMetadata.getSourceAsMap();
+        String modelIdMappingString = String.join("model_id=", task.getModelId());
+        // If modelId is present, parse map to field.
+        if (mappingMetadataSourceMap.toString().contains(modelIdMappingString)) {
+            for (Map.Entry<String, Object> sourceEntry : mappingMetadataSourceMap.entrySet()) {
+                if (sourceEntry.getKey() != null && sourceEntry.getKey().equals(PROPERTIES) && sourceEntry.getValue() instanceof Map) {
+                    Map<String, Object> fieldsMap = (Map<String, Object>) sourceEntry.getValue();
+                    indicesUsingModel = parseFieldsMap(fieldsMap, task, indicesUsingModel, index);
+                }
+            }
+        }
+        return indicesUsingModel;
+    }
+
+    private static List<String> parseFieldsMap(
+        Map<String, Object> fieldsMap,
+        UpdateModelGraveyardTask task,
+        List<String> indicesUsingModel,
+        String index
```

Comment:
Is this necessary?

Reply:
I guess I could return true and add the index higher up, but I use it on line 232